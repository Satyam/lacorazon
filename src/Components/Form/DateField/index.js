import React, { useContext, useEffect, useState } from 'react';
import { FormGroup, Label, FormFeedback, FormText, Col } from 'reactstrap';
import { FormikConsumer, ErrorMessage } from 'formik';
import invariant from 'invariant';
import DatePicker from 'react-datepicker';

import 'react-datepicker/dist/react-datepicker.css';
import classNames from 'classnames';

function useFormik() {
  return useContext(FormikConsumer._context);
}

let counter = 0;

export default function DateField({
  name,
  label,
  help,
  className,
  id,
  ...rest
}) {
  invariant(name, 'DateField: name argument is mandatory');

  const {
    values,
    errors,
    touched,
    registerField,
    unregisterField,
    setFieldTouched,
    setFieldValue,
    validationSchema,
  } = useFormik();

  useEffect(() => {
    registerField(name, {
      props: {
        name,
        label,
        help,
        className,
        id,
        ...rest,
      },
    });
    return () =>
      unregisterField(name, {
        props: {
          name,
          label,
          help,
          className,
          id,
          ...rest,
        },
      });
  });

  const [actualId] = useState(id || `F_DF_${counter}`);

  counter = (counter + 1) % Number.MAX_SAFE_INTEGER;

  const invalid = errors[name] && touched[name];

  let actualMin = rest.minDate;
  let actualMax = rest.maxDate;

  if (validationSchema) {
    const tests = validationSchema.fields[name].tests;
    if (!actualMin) {
      const minTest = tests.filter(t => t.TEST_NAME === 'min')[0];
      if (minTest) {
        actualMin = minTest.TEST.params.min;
      }
    }
    if (!actualMax) {
      const maxTest = tests.filter(t => t.TEST_NAME === 'max')[0];
      if (maxTest) {
        actualMax = maxTest.TEST.params.max;
      }
    }
  }
  return (
    <FormGroup row>
      <Label for={actualId} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <DatePicker
          className={classNames('form-control', className, {
            'is-invalid': invalid,
          })}
          name={name}
          id={actualId}
          minDate={actualMin}
          maxDate={actualMax}
          onChange={value => setFieldValue(name, value)}
          onBlur={() => setFieldTouched(name, true)}
          selected={values[name]}
          {...rest}
        />
        {help && <FormText>{help}</FormText>}
        <ErrorMessage
          name={name}
          component={FormFeedback}
          style={{ display: invalid ? 'block' : 'none' }}
        />
      </Col>
    </FormGroup>
  );
}
