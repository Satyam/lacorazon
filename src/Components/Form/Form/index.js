import React from 'react';
import { Form as BSForm, Alert } from 'reactstrap';
import { Formik, Form as KForm } from 'formik';

export default function Form({
  schema,
  values,
  onSubmit,
  children,
  enableReinitialize = true,
  isInitialValid = false,
  onReset,
  validate,
  validateOnBlur = true,
  validateOnChange = true,
  ...rest
}) {
  return (
    <Formik
      validationSchema={schema}
      initialValues={schema ? Object.assign(schema.default(), values) : values}
      onSubmit={(values, formik) => {
        const result = onSubmit(schema ? schema.cast(values) : values, formik);
        if (result && typeof result.then === 'function') {
          result.then(
            () => {
              formik.setSubmitting(false);
            },
            err => {
              formik.setFieldError('*', err);
            }
          );
        }
      }}
      enableReinitialize={enableReinitialize}
      isInitialValid={isInitialValid}
      onReset={onReset}
      validate={validate}
      validateOnBlur={validateOnBlur}
      validateOnChange={validateOnChange}
    >
      {({ errors }) => (
        <BSForm tag={KForm} {...rest}>
          {errors['*'] && <Alert color="danger">{errors['*']}</Alert>}
          {children}
        </BSForm>
      )}
    </Formik>
  );
}
