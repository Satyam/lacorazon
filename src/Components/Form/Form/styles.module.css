.form-border {
  border: #ced4da solid thin;
  padding: 0.5rem;
  margin: 0.5rem;
  border-radius: 0.3rem;
}
