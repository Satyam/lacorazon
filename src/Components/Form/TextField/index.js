import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { FormGroup, Label, FormFeedback, FormText, Col } from 'reactstrap';
import { Field as KField, FormikConsumer, ErrorMessage } from 'formik';
import classNames from 'classnames';
import invariant from 'invariant';

function useFormik() {
  return useContext(FormikConsumer._context);
}

let counter = 0;
/**
 * Produces a labeled input box within form
 */
export default function TextField({
  name,
  label,
  id,
  rows,
  help,
  validate,
  ...rest
}) {
  invariant(name, 'TextField: name argument is mandatory');

  const { errors, touched, validationSchema } = useFormik();
  const invalid = errors[name] && touched[name];
  const [actualId] = useState(id || `F_TF_${counter}`);
  counter = (counter + 1) % Number.MAX_SAFE_INTEGER;

  return (
    <FormGroup row>
      <Label for={actualId} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <KField
          component={rows ? 'textarea' : 'input'}
          className={classNames('form-control', {
            'is-invalid': invalid,
          })}
          rows={rows}
          name={name}
          id={actualId}
          validate={
            validate
              ? value =>
                  validate(
                    validationSchema
                      ? validationSchema.fields[name].cast(value)
                      : value
                  )
              : void 0
          }
          {...rest}
        />
        {help && <FormText>{help}</FormText>}
        <ErrorMessage name={name} component={FormFeedback} />
      </Col>
    </FormGroup>
  );
}

TextField.propTypes = {
  /**
   * Name of the field this component is to show or edit.
   * It should match the key of one of the values in the
   * enclosing [`Form`](#form) `values` property.
   */
  name: PropTypes.string.isRequired,
  /**
   * The label to be shown to the user along the input box
   */
  label: PropTypes.string,
  /**
   * An optional `id` attribute.
   * If none is provided, the component will generate a unique one
   * so the `<label for="inputId">` element can match it
   */
  id: PropTypes.string,
  /**
   * If provided, a `<textarea>` will be produced with as many rows
   * as given, instead of a regular `<input>`box
   */
  rows: PropTypes.number,
  /**
   * An optional help text to be shown below the input field
   */
  help: PropTypes.string,
  /**
   * A validation function.
   * It will be called with the value of the field and
   * it should return a text with an error message, if any,
   * or nothing if Ok.
   */
  validate: PropTypes.func,
  /**
   * Any other properties will be passed on to the `<input>` or `<textarea>` elements
   */
  '...rest': PropTypes.object,
};
