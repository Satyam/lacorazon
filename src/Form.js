import React, { useContext } from 'react';
import {
  Form as BSForm,
  FormGroup,
  Label,
  FormFeedback,
  FormText,
  Alert,
  Col,
  Button
} from 'reactstrap';
import { Formik, Form as KForm, Field as KField, FormikConsumer } from 'formik';

function useFormik() {
  return useContext(FormikConsumer._context);
}

export function Form({
  initialValues,
  schema,
  values,
  onSubmit,
  children,
  ...rest
}) {
  return (
    <Formik
      enableReinitialize={true}
      validationSchema={schema}
      initialValues={Object.assign(schema.default(), values)}
      onSubmit={(values, formik) => onSubmit(schema.cast(values), formik)}
      {...rest}
    >
      {({ errors }) => (
        <BSForm tag={KForm}>
          {errors['*'] && <Alert color="danger">{errors['*']}</Alert>}
          {children}
        </BSForm>
      )}
    </Formik>
  );
}

export function TextField({ name, label, rows, help, validate, ...rest }) {
  const { errors, touched, validationSchema } = useFormik();
  const error = errors[name];

  return (
    <FormGroup row>
      <Label for={name} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <KField
          component={rows ? 'textarea' : 'input'}
          className="form-control"
          rows={rows}
          name={name}
          id={name}
          invalid={error && touched[name]}
          validate={
            validate
              ? value => validate(validationSchema.fields[name].cast(value))
              : void 0
          }
          {...rest}
        />
        <FormFeedback>{error}</FormFeedback>
        {help && <FormText>{help}</FormText>}
      </Col>
    </FormGroup>
  );
}

export function SubmitButton({ component: Component = Button, ...rest }) {
  const { isSubmitting, isValid, dirty } = useFormik();
  return (
    <Component
      type="submit"
      disabled={isSubmitting || !isValid || !dirty}
      {...rest}
    />
  );
}
