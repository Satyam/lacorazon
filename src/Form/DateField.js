import React, { useContext, useEffect } from 'react';
import { FormGroup, Label, FormFeedback, FormText, Col } from 'reactstrap';
import { FormikConsumer, ErrorMessage } from 'formik';

import DatePicker from 'react-datepicker';

import 'react-datepicker/dist/react-datepicker.css';
import classNames from 'classnames';

function useFormik() {
  return useContext(FormikConsumer._context);
}

export default function DateField({
  name,
  label,
  help,
  validate,
  className,
  ...rest
}) {
  if (process.env.NODE_ENV !== 'production' && !name) {
    throw new Error('TextField: name argument is mandatory');
  }
  const {
    values,
    errors,
    touched,
    validationSchema,
    registerField,
    unregisterField,
    setTouched,
    setFieldValue,
    validateField,
    validateOnBlur,
    validateOnChange
  } = useFormik();

  useEffect(() => {
    registerField(name, this);
    return () => unregisterField(name, this);
  });

  const invalid = errors[name] && touched[name];
  return (
    <FormGroup row>
      <Label for={name} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <DatePicker
          className={classNames('form-control', className, {
            'is-invalid': invalid
          })}
          name={name}
          id={name}
          onChange={value => {
            setFieldValue(name, value);
            if (validateOnChange) {
              validateField(name);
            }
          }}
          onBlur={() => {
            setTouched(name, true);
            if (validateOnBlur) {
              validateField(name);
            }
          }}
          selected={values[name]}
          validate={
            validate
              ? value =>
                  validate(
                    validationSchema
                      ? validationSchema.fields[name].cast(value)
                      : value
                  )
              : void 0
          }
          {...rest}
        />
        {help && <FormText>{help}</FormText>}
        <ErrorMessage
          name={name}
          component={FormFeedback}
          style={{ display: invalid ? 'block' : 'none' }}
        />
      </Col>
    </FormGroup>
  );
}
