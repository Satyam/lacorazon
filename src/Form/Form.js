import React from 'react';
import { Form as BSForm, Alert } from 'reactstrap';
import { Formik, Form as KForm } from 'formik';

export default function Form({
  initialValues,
  schema,
  values,
  onSubmit,
  children,
  ...rest
}) {
  return (
    <Formik
      enableReinitialize={true}
      validationSchema={schema}
      initialValues={Object.assign(schema.default(), values)}
      onSubmit={(values, formik) => onSubmit(schema.cast(values), formik)}
      {...rest}
    >
      {({ errors }) => (
        <BSForm tag={KForm}>
          {errors['*'] && <Alert color="danger">{errors['*']}</Alert>}
          {children}
        </BSForm>
      )}
    </Formik>
  );
}
