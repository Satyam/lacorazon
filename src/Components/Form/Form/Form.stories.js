import React, { useContext, useState } from 'react';

import { storiesOf } from '@storybook/react';

import { FormikConsumer } from 'formik';

import Form from '.';
import styles from './styles.module.css';

function useFormik() {
  return useContext(FormikConsumer._context);
}

function ForceValidate() {
  const [done, setDone] = useState(false);
  const { validateForm } = useFormik();
  if (done) return null;
  validateForm();
  setDone(true);
  return null;
}

function ForceSubmit() {
  const [done, setDone] = useState(false);
  const { submitForm } = useFormik();
  if (done) return null;
  submitForm();
  setDone(true);
  return null;
}

storiesOf('Form/Form', module)
  .add('no attributes', () => <Form />)
  .add('with some child', () => <Form>Hello</Form>)
  .add('with some border and child', () => (
    <Form className={styles['form-border']}>Hello</Form>
  ))
  .add('with error', () => (
    <Form
      isInitialValid={false}
      validate={() => ({ '*': 'some made-up error' })}
      values={{}}
    >
      <ForceValidate />
    </Form>
  ))
  .add('with rejected promise validation', () => (
    <Form
      isInitialValid={false}
      onSubmit={() => Promise.reject('some made-up error')}
      values={{}}
    >
      <ForceSubmit />
    </Form>
  ));
