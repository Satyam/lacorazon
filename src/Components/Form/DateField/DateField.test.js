import React from 'react';
import { render, fireEvent, cleanup, act } from '@testing-library/react';
import * as Yup from 'yup';

import Form from '../Form';
import DateField from './';

afterEach(cleanup);

class ErrorBoundary extends React.PureComponent {
  state = { hasError: false };

  static getDerivedStateFromError(err) {
    // Update state so the next render will show the fallback UI.
    return { hasError: err.message };
  }

  // componentDidCatch(err, info) {
  //   this.setState({ hasError: err.message });
  // }

  render() {
    return this.state.hasError || React.Children.only(this.props.children);
  }
}

describe('Form/DateField', () => {
  it('should throw with no props as name argument is mandatory', () => {
    const e = console.error;
    console.error = msg => {};
    const { container } = render(
      <ErrorBoundary>
        <Form>
          <DateField />
        </Form>
      </ErrorBoundary>
    );

    expect(container.innerHTML).toMatchSnapshot();
    console.error = e;
  });
  it('should throw with any extra property but name as argument is mandatory', () => {
    const e = console.error;
    console.error = msg => {};
    const { container } = render(
      <ErrorBoundary>
        <Form>
          <DateField label="some label" value="Some value" />
        </Form>
      </ErrorBoundary>
    );
    expect(container.innerHTML).toMatchSnapshot();
    console.error = e;
  });

  it('should validate on field change', () => {
    const validate = jest.fn(() => '');
    const { getByLabelText } = render(
      <Form values={{ one: new Date(2019, 8, 7) }}>
        <DateField label="one" name="one" validate={validate} />
      </Form>
    );
    fireEvent.change(getByLabelText('one'), {
      target: { value: new Date(2019, 2, 2) },
    });
    expect(validate.mock.calls).toEqual([[new Date(2019, 2, 2)]]);
  });

  it('should validate on field blur', () => {
    const validate = jest.fn(() => '');
    const { getByLabelText } = render(
      <Form values={{ one: new Date(2019, 8, 7) }}>
        <DateField label="one" name="one" validate={validate} />
      </Form>
    );

    fireEvent.blur(getByLabelText('one'));
    expect(validate.mock.calls).toEqual([[new Date(2019, 8, 7)]]);
  });

  it('should generate an id when no id provided', () => {
    const { getByLabelText } = render(
      <Form values={{ one: new Date(2019, 8, 7) }}>
        <DateField label="one" name="one" />
      </Form>
    );
    expect(getByLabelText('one').id).toMatch(/^F_DF_\d+$/);
  });

  it('should respect the id provided', () => {
    const { getByLabelText } = render(
      <Form values={{ one: new Date(2019, 8, 7) }}>
        <DateField label="one" name="one" id="abcd" />
      </Form>
    );
    expect(getByLabelText('one').id).toBe('abcd');
  });

  it('should take values from the schema', () => {
    const schema = Yup.object().shape({
      one: Yup.date().default(new Date(2019, 8, 7)),
    });
    const validate = jest.fn(() => '');
    const { getByLabelText } = render(
      <Form schema={schema}>
        <DateField label="one" name="one" validate={validate} />
      </Form>
    );
    fireEvent.blur(getByLabelText('one'));
    expect(validate.mock.calls).toEqual([[new Date(2019, 8, 7)]]);
  });
  it('should reject values below the min in the schema', done => {
    // since the out-of-range dates are not enabled, they can't be clicked
    const schema = Yup.object().shape({
      one: Yup.date()
        .min(new Date(2019, 8, 10))
        .default(new Date(2019, 8, 20)),
    });
    let wrapper;
    act(() => {
      wrapper = render(
        <Form schema={schema}>
          <DateField label="one" name="one" />
        </Form>
      );
    });
    fireEvent.click(wrapper.getByLabelText('one'));
    fireEvent.click(wrapper.getByText('6'));
    fireEvent.blur(wrapper.getByLabelText('one'));
    setTimeout(() => {
      expect(wrapper.getByLabelText('one')).toHaveClass('is-invalid');
      expect(
        wrapper.container.querySelector('.invalid-feedback')
      ).toBeVisible();
      done();
    }, 1);
  });
});
