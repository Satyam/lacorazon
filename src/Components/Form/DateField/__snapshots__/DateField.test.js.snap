// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form/DateField should throw with any extra property but name as argument is mandatory 1`] = `"DateField: name argument is mandatory"`;

exports[`Form/DateField should throw with no props as name argument is mandatory 1`] = `"DateField: name argument is mandatory"`;
